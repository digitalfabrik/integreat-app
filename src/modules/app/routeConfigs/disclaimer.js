// @flow

import RouteConfig from './RouteConfig'
import type { Dispatch, GetState, Route } from 'redux-first-router'
import fetchData from '../fetchData'
import disclaimerEndpoint from '../../endpoint/endpoints/disclaimer'
import type { GetLanguageChangePathParamsType, GetPageTitleParamsType } from './RouteConfig'

type DisclaimerRouteParamsType = {|city: string, language: string|}

export const DISCLAIMER_ROUTE = 'DISCLAIMER'

const getDisclaimerPath = ({city, language}: DisclaimerRouteParamsType): string =>
  `/${city}/${language}/disclaimer`

/**
 * DisclaimerRoute (for city specific disclaimers), matches /augsburg/de/disclaimer
 * @type {{path: string, thunk: function(Dispatch, GetState)}}
 */
const disclaimerRoute: Route = {
  path: '/:city/:language/disclaimer',
  thunk: async (dispatch: Dispatch, getState: GetState) => {
    const state = getState()
    const {city, language} = state.location.payload

    await fetchData(disclaimerEndpoint, dispatch, state.disclaimer, {city, language})
  }
}

const getLanguageChangePath = ({location, language}: GetLanguageChangePathParamsType) =>
  getDisclaimerPath({city: location.payload.city, language})

const getPageTitle = ({t, cityName}: GetPageTitleParamsType) =>
  `${t('pageTitles.disclaimer')} - ${cityName}`

class DisclaimerRouteConfig extends RouteConfig<DisclaimerRouteParamsType> {
  constructor () {
    super({
      name: DISCLAIMER_ROUTE,
      route: disclaimerRoute,
      getRoutePath: getDisclaimerPath,
      getLanguageChangePath,
      getPageTitle
    })
  }
}

export default DisclaimerRouteConfig
