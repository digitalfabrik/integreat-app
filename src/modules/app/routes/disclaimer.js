// @flow

import disclaimerEndpoint from '../../endpoint/endpoints/disclaimer'
import { createAction } from 'redux-actions'

import type { Dispatch, GetState, Action } from 'redux-first-router'
import CityModel from '../../endpoint/models/CityModel'
import PageModel from '../../endpoint/models/PageModel'
import DisclaimerPage from '../../../routes/disclaimer/containers/DisclaimerPage'
import React from 'react'
import Route from './Route'
import Payload from '../../endpoint/Payload'
import type { AllPayloadsType } from './types'
import fetchData from '../fetchData'

type RequiredPayloadType = {|disclaimer: Payload<PageModel>, cities: Payload<Array<CityModel>>|}
type RouteParamsType = {|city: string, language: string|}

export const DISCLAIMER_ROUTE = 'DISCLAIMER'

export const goToDisclaimer = (city: string, language: string): Action =>
  createAction<string, { city: string, language: string }>(DISCLAIMER_ROUTE)({
    city,
    language
  })

const getRoutePath = ({city, language}: RouteParamsType): string => `/${city}/${language}/disclaimer`

const renderDisclaimerPage = ({disclaimer, cities}: RequiredPayloadType) =>
  <DisclaimerPage disclaimer={disclaimer} cities={cities} />

const getRequiredPayloads = (payloads: AllPayloadsType) =>
  ({disclaimer: payloads.disclaimerPayload, cities: payloads.citiesPayload})

/**
 * DisclaimerRoute (for city specific disclaimers), matches /augsburg/de/disclaimer
 * @type {{path: string, thunk: function(Dispatch, GetState)}}
 */
const disclaimerRoute = {
  path: '/:city/:language/disclaimer',
  thunk: async (dispatch: Dispatch, getState: GetState) => {
    const state = getState()
    const {city, language} = state.location.payload

    await fetchData(disclaimerEndpoint, dispatch, state.disclaimer, {city, language})
  }
}

export default new Route<RequiredPayloadType, RouteParamsType>({
  name: DISCLAIMER_ROUTE,
  getRoutePath,
  renderPage: renderDisclaimerPage,
  route: disclaimerRoute,
  getRequiredPayloads
})
