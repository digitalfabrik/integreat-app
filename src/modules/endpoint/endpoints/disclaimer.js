// @flow

import { apiUrl } from '../constants'
import DisclaimerModel from '../models/DisclaimerModel'
import { isEmpty } from 'lodash/lang'
import EndpointBuilder from '../EndpointBuilder'
import ParamMissingError from '../errors/ParamMissingError'
import type { EndpointParams } from '../../../flowTypes'
import moment from 'moment/moment'

const DISCLAIMER_ENDPOINT_NAME = 'disclaimer'

export default new EndpointBuilder(DISCLAIMER_ENDPOINT_NAME)
  .withParamsToUrlMapper((params: EndpointParams): string => {
    if (!params.city) {
      throw new ParamMissingError(DISCLAIMER_ENDPOINT_NAME, 'city')
    }
    if (!params.language) {
      throw new ParamMissingError(DISCLAIMER_ENDPOINT_NAME, 'language')
    }
    return `${apiUrl}/${params.city}/${params.language}/wp-json/extensions/v3/disclaimer`
  })
  .withMapper((json: any): DisclaimerModel => {
    if (isEmpty(json)) {
      throw new Error('disclaimer:notAvailable')
    }

    return new DisclaimerModel({
      id: json.id,
      title: json.title,
      content: json.content,
      lastUpdate: moment(json.modified_gmt)
    })
  })
  .build()
