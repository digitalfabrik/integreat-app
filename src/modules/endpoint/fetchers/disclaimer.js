import { apiUrl } from '../constants'
import DisclaimerModel from '../models/DisclaimerModel'
import { isEmpty } from 'lodash/lang'

export const urlMapper = params => `${apiUrl}/${params.location}/${params.language}/wp-json/extensions/v0/modified_content/disclaimer?since=1970-01-01T00:00:00Z`

const mapper = json => {
  if (isEmpty(json)) {
    throw new Error('disclaimer:notAvailable')
  }

  const disclaimers = json
    .filter(disclaimer => disclaimer.status === 'publish')
    .map(disclaimer => {
      return new DisclaimerModel({
        id: disclaimer.id,
        title: disclaimer.title,
        content: disclaimer.content
      })
    })

  if (disclaimers.length !== 1) {
    throw new Error('There must be exactly one disclaimer!')
  }

  return disclaimers[0]
}

const fetcher = async params =>
  fetch(urlMapper(params))
    .then(json => mapper(json))

export default fetcher
