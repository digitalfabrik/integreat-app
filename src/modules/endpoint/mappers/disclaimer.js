import DisclaimerModel from '../models/DisclaimerModel'
import { isEmpty } from 'lodash/lang'

function disclaimerMapper (json) {
  if (isEmpty(json)) {
    throw new Error('disclaimer:notAvailable')
  }

  const disclaimers = json
    .filter(disclaimer => disclaimer.status === 'publish')
    .map(disclaimer => {
      return new DisclaimerModel({
        id: disclaimer.id,
        title: disclaimer.title,
        content: disclaimer.content
      })
    })

  if (disclaimers.length !== 1) {
    throw new Error('There must be exactly one disclaimer!')
  }

  return disclaimers[0]
}

export default disclaimerMapper
