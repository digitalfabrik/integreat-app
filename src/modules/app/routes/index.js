// @flow

import categoriesRoute from './categories'
import eventsRoute from './events'
import { EXTRAS_ROUTE, extrasRoute, getExtrasPath, renderExtrasPage } from './extras'
import { DISCLAIMER_ROUTE, disclaimerRoute, getDisclaimerPath, renderDisclaimerPage } from './disclaimer'
import poisRoute from './pois'
import { getSprungbrettExtraPath, renderSprungbrettPage, SPRUNGBRETT_ROUTE, sprungbrettRoute } from './sprungbrett'
import { getWohnenExtraPath, renderWohnenPage, WOHNEN_ROUTE, wohnenRoute } from './wohnen'
import { getSearchPath, renderSearchPage, SEARCH_ROUTE, searchRoute } from './search'
import { MAIN_DISCLAIMER_ROUTE, mainDisclaimerRoute, renderMainDisclaimerPage } from './mainDisclaimer'
import { I18N_REDIRECT_ROUTE, i18nRedirectRoute, renderI18nPage } from './i18nRedirect'
import { LANDING_ROUTE, landingRoute, renderLandingPage } from './landing'
import type { Route as RouterRouteType } from 'redux-first-router'
import Route from './Route'

export const requiredPayloads = {
  [EXTRAS_ROUTE]: (payloads: AllPayloadsType) =>
    ({extras: payloads.extrasPayload, cities: payloads.citiesPayload}),
  [DISCLAIMER_ROUTE]: (payloads: AllPayloadsType) =>
    ({disclaimer: payloads.disclaimerPayload, cities: payloads.citiesPayload}),
  [SPRUNGBRETT_ROUTE]: (payloads: AllPayloadsType) => ({
    extras: payloads.extrasPayload, cities: payloads.citiesPayload, sprungbrettJobs: payloads.sprungbrettJobsPayload}),
  [WOHNEN_ROUTE]: (payloads: AllPayloadsType) =>
    ({extras: payloads.extrasPayload, cities: payloads.citiesPayload, offers: payloads.wohnenPayload}),
  [SEARCH_ROUTE]: (payloads: AllPayloadsType) =>
    ({categories: payloads.categoriesPayload, cities: payloads.citiesPayload}),
  [MAIN_DISCLAIMER_ROUTE]: () => ({}),
  [I18N_REDIRECT_ROUTE]: (payloads: AllPayloadsType) => ({cities: payloads.citiesPayload}),
  [LANDING_ROUTE]: (payloads: AllPayloadsType) => ({cities: payloads.citiesPayload})
}

// this maps all goToRoute actions to the right routes (except from the NOT_FOUND route)
// the order is important, routes declared first are served first, so i.e. if you put the mainDisclaimer route after
// the i18nRedirect route, "/disclaimer" also matches the i18nRedirect route
export const routesMap: {[string]: RouterRouteType} = {
  [MAIN_DISCLAIMER_ROUTE]: mainDisclaimerRoute,
  [I18N_REDIRECT_ROUTE]: i18nRedirectRoute,
  [LANDING_ROUTE]: landingRoute,
  [eventsRoute.name]: eventsRoute,
  [SPRUNGBRETT_ROUTE]: sprungbrettRoute,
  [WOHNEN_ROUTE]: wohnenRoute,
  [EXTRAS_ROUTE]: extrasRoute,
  [DISCLAIMER_ROUTE]: disclaimerRoute,
  [SEARCH_ROUTE]: searchRoute,
  [poisRoute.name]: poisRoute,
  [categoriesRoute.name]: categoriesRoute
}

const routes: {[string]: Route} = {
  [categoriesRoute.name]: categoriesRoute,
  [eventsRoute.name]: eventsRoute,
  [EXTRAS_ROUTE]: renderExtrasPage,
  [DISCLAIMER_ROUTE]: renderDisclaimerPage,
  [poisRoute.name]: poisRoute,
  [SPRUNGBRETT_ROUTE]: renderSprungbrettPage,
  [WOHNEN_ROUTE]: renderWohnenPage,
  [SEARCH_ROUTE]: renderSearchPage,
  [MAIN_DISCLAIMER_ROUTE]: renderMainDisclaimerPage,
  [I18N_REDIRECT_ROUTE]: renderI18nPage,
  [LANDING_ROUTE]: renderLandingPage
}

export default routes
