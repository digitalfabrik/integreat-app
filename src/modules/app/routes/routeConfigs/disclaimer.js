// @flow

import React from 'react'
import { DISCLAIMER_ROUTE, getDisclaimerPath } from '../disclaimer'
import RouteConfig from './RouteConfig'
import type { AllPayloadsType, GetLanguageChangePathParamsType, GetPageTitleParamsType } from '../types'
import Payload from '../../../endpoint/Payload'
import PageModel from '../../../endpoint/models/PageModel'
import DisclaimerPage from '../../../../routes/disclaimer/containers/DisclaimerPage'

type RequiredPayloadType = {|disclaimer: Payload<PageModel>|}

const renderDisclaimerPage = ({disclaimer}: RequiredPayloadType) =>
  <DisclaimerPage disclaimer={disclaimer.data} />

const getRequiredPayloads = (payloads: AllPayloadsType) =>
  ({disclaimer: payloads.disclaimerPayload})

const getLanguageChangePath = ({location, language}: GetLanguageChangePathParamsType) =>
  getDisclaimerPath({city: location.payload.city, language})

const getPageTitle = ({t, cityName}: GetPageTitleParamsType) =>
  `${t('pageTitles.disclaimer')} - ${cityName}`

const disclaimerRouteConfig: RouteConfig<RequiredPayloadType> = new RouteConfig({
  name: DISCLAIMER_ROUTE,
  renderPage: renderDisclaimerPage,
  getRequiredPayloads,
  getLanguageChangePath,
  getPageTitle
})

export default disclaimerRouteConfig
