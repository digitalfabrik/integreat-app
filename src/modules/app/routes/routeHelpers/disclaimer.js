// @flow

import React from 'react'
import { getDisclaimerPath } from '../disclaimer'
import Route from './RouteHelper'
import type { AllPayloadsType, GetLanguageChangePathParamsType, GetPageTitleParamsType } from '../types'
import Payload from '../../../endpoint/Payload'
import PageModel from '../../../endpoint/models/PageModel'
import DisclaimerPage from '../../../../routes/disclaimer/containers/DisclaimerPage'

type RequiredPayloadType = {|disclaimer: Payload<PageModel>|}

const renderDisclaimerPage = ({disclaimer}: RequiredPayloadType) =>
  <DisclaimerPage disclaimer={disclaimer.data} />

const getRequiredPayloads = (payloads: AllPayloadsType) =>
  ({disclaimer: payloads.disclaimerPayload})

const getLanguageChangePath = ({location, language}: GetLanguageChangePathParamsType) =>
  getDisclaimerPath({city: location.payload.city, language})

const getPageTitle = ({t, cityName}: GetPageTitleParamsType) =>
  `${t('disclaimerPageTitle')} - ${cityName}`

const disclaimerRouteHelper: Route<RequiredPayloadType> = new Route({
  renderPage: renderDisclaimerPage,
  getRequiredPayloads,
  getLanguageChangePath,
  getPageTitle
})

export default disclaimerRouteHelper
