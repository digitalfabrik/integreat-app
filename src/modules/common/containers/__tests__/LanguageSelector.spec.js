import React from 'react'
import { shallow } from 'enzyme'

import ConnectedLanguageSelector, { LanguageSelector } from '../LanguageSelector'
import LanguageModel from '../../../endpoint/models/LanguageModel'
import EventModel from '../../../endpoint/models/EventModel'
import { DISCLAIMER_ROUTE } from '../../../app/routes/disclaimer'
import CategoriesMapModel from '../../../endpoint/models/CategoriesMapModel'
import CategoryModel from '../../../endpoint/models/CategoryModel'
import configureMockStore from 'redux-mock-store'

describe('LanguageSelector', () => {
  const city = 'augsburg'
  const language = 'en'
  const languages = [
    new LanguageModel('de', 'Deutsch'),
    new LanguageModel('en', 'English'),
    new LanguageModel('ar', 'Arabic')
  ]

  const events = [
    new EventModel({
      id: '1234',
      title: 'nulltes Event',
      availableLanguages: {
        de: '1',
        en: '2'
      }
    })]

  const categoryModels = [
    new CategoryModel({
      id: 3650,
      path: '/augsburg/en/welcome',
      title: 'Welcome',
      content: '',
      parentPath: '/augsburg/en',
      order: 75,
      availableLanguages: new Map([['de', '/augsburg/de/willkommen']]),
      thumbnail: 'https://cms.integreat-apâ€¦/03/Hotline-150x150.png'
    })]

  const categories = new CategoriesMapModel(categoryModels)

  it('should render a HeaderLanguageSelectorItem if it is a header action item', () => {
    const location = {
      pathname: '/augsburg/en/disclaimer',
      type: DISCLAIMER_ROUTE,
      payload: {city, language}
    }

    const languageSelector = shallow(
      <LanguageSelector categories={categories}
                        events={events}
                        languages={languages}
                        location={location}
                        isHeaderActionItem
                        t={key => key} />
    )

    expect(languageSelector).toMatchSnapshot()
  })

  it('should render a normal Selector if it is not a header action item', () => {
    const location = {
      pathname: '/augsburg/en/disclaimer',
      type: DISCLAIMER_ROUTE,
      payload: {city, language}
    }

    const languageSelector = shallow(
      <LanguageSelector categories={categories}
                        events={events}
                        languages={languages}
                        location={location}
                        isHeaderActionItem={false}
                        t={key => key} />
    )

    expect(languageSelector).toMatchSnapshot()
  })

  it('should map state to props', () => {
    const location = {type: 'DISCLAIMER', payload: {city, language}, pathname: '/augsburg/de/disclaimer'}

    const mockStore = configureMockStore()
    const store = mockStore({
      location: location,
      languages: {data: languages},
      categories: {data: categories},
      events: {data: events}
    })

    const languageSelector = shallow(
      <ConnectedLanguageSelector isHeaderActionItem store={store} />
    )

    expect(languageSelector.props()).toMatchObject({
      languages,
      location,
      events,
      categories,
      isHeaderActionItem: true
    })
  })
})
