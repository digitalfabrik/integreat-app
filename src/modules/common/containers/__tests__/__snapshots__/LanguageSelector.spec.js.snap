// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSelector should add vertical class name if verticalLayout is true 1`] = `
<Selector
  activeItemCode="de"
  closeDropDownCallback={[Function]}
  items={
    Array [
      SelectorItemModel {
        "_code": "de",
        "_name": "Deutsch",
        "_path": "test/url/de",
      },
      SelectorItemModel {
        "_code": "en",
        "_name": "English",
        "_path": "test/url/en",
      },
      SelectorItemModel {
        "_code": "ar",
        "_name": "Arabic",
        "_path": "test/url/ar",
      },
    ]
  }
  verticalLayout={true}
/>
`;

exports[`LanguageSelector should match snapshot 1`] = `
<Selector
  activeItemCode="de"
  closeDropDownCallback={[Function]}
  items={
    Array [
      SelectorItemModel {
        "_code": "de",
        "_name": "Deutsch",
        "_path": "test/url/de",
      },
      SelectorItemModel {
        "_code": "en",
        "_name": "English",
        "_path": "test/url/en",
      },
      SelectorItemModel {
        "_code": "ar",
        "_name": "Arabic",
        "_path": "test/url/ar",
      },
    ]
  }
/>
`;
